File: frontend/src/domain/entities/Order.js
import { Address } from './User.js';

// Order Item Entity
export class OrderItem {
    constructor(data = {}) {
        this.productId = data.product_id || data.productId || '';
        this.productName = data.product_name || data.productName || '';
        this.productImage = data.image_url || data.product_image || data.productImage || '';
        this.color = data.color || '';
        this.size = data.size || '';
        this.quantity = data.quantity || 1;
        this.price = data.price || 0;
    }

    // Get subtotal
    getSubtotal() {
        return this.price * this.quantity;
    }
}

// Order Entity
export class Order {
    constructor(data = {}) {
        this.id = data.id || data._id || '';
        this.orderId = data.order_id || data.orderId || '';
        this.userId = data.user_id || data.userId || '';
        this.items = (data.items || []).map(item => new OrderItem(item));
        this.totalAmount = data.total_amount || data.totalAmount || 0;
        this.status = data.status || 'Pending';
        this.shippingMethod = data.shipping_method || data.shippingMethod || '';
        this.shippingStatus = data.shipping_status || data.shippingStatus || 'Pending';
        this.shippingAddress = new Address(data.shipping_address || data.shippingAddress || {});
        this.envioPackId = data.enviopack_id || data.envioPackId || '';
        this.paymentUrl = data.payment_url || data.paymentUrl || '';
        this.createdAt = data.created_at || data.createdAt;
        this.updatedAt = data.updated_at || data.updatedAt;
    }

    // Calculate total from items
    calculateTotal() {
        return this.items.reduce((total, item) => total + item.getSubtotal(), 0);
    }

    // Check if order is paid
    isPaid() {
        return this.status === 'Paid';
    }

    // Check if order is shipped
    isShipped() {
        return this.shippingStatus === 'Shipped' || this.shippingStatus === 'Delivered';
    }

    // Check if order is pending shipment
    isPendingShipment() {
        return this.isPaid() && this.shippingStatus === 'Pending';
    }

    // Check if shipping is via WhatsApp
    isWhatsAppShipping() {
        return this.shippingMethod === 'Whatsapp';
    }

    // Check if shipping is via Send
    isSendShipping() {
        return this.shippingMethod === 'Send';
    }

    // Validate order
    validate() {
        const errors = [];

        if (!this.userId) {
            errors.push('User ID is required');
        }

        if (!this.items || this.items.length === 0) {
            errors.push('At least one item is required');
        }

        if (!this.shippingMethod) {
            errors.push('Shipping method is required');
        }

        if (this.totalAmount <= 0) {
            errors.push('Total amount must be greater than 0');
        }

        const addressValidation = this.shippingAddress.validate();
        if (!addressValidation.isValid) {
            errors.push(...addressValidation.errors.map(e => `Shipping address: ${e}`));
        }

        return {
            isValid: errors.length === 0,
            errors
        };
    }
}

export default Order;

File: payments-service/internal/domain/order.go
package domain

import (
	"errors"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// OrderStatus represents the status of an order
type OrderStatus string

const (
	StatusPending   OrderStatus = "Pending"
	StatusPaid      OrderStatus = "Paid"
	StatusShipped   OrderStatus = "Shipped"
	StatusCancelled OrderStatus = "Cancelled"
)

const (
	ShippingStatusPending   string = "Pending"
	ShippingStatusShipped   string = "Shipped"
	ShippingStatusDelivered string = "Delivered"
	ShippingStatusCancelled string = "Cancelled"
)

const (
	ShippingMethodSend     string = "Send"
	ShippingMethodWhatsapp string = "Whatsapp"
)

// OrderItem represents a product in an order
type OrderItem struct {
	ProductID   string  `bson:"product_id" json:"product_id"`
	ProductName string  `bson:"product_name" json:"product_name"`
	ImageURL    string  `bson:"image_url" json:"image_url"`
	Color       string  `bson:"color" json:"color"`
	Size        string  `bson:"size" json:"size"`
	Quantity    int     `bson:"quantity" json:"quantity"`
	Price       float64 `bson:"price" json:"price"`
	Weight      float64 `bson:"weight" json:"weight"`
}

type Address struct {
	Street    string `bson:"street" json:"street"`
	Number    string `bson:"number" json:"number"`
	Floor     string `bson:"floor" json:"floor"`
	Apartment string `bson:"apartment" json:"apartment"`
	City      string `bson:"city" json:"city"`
	Province  string `bson:"province" json:"province"`
	Country   string `bson:"country" json:"country"`
	ZipCode   string `bson:"zip_code" json:"zip_code"`
}

// Order represents an order entity
type Order struct {
	ID                  primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	OrderID             string             `bson:"order_id" json:"order_id"`
	UserID              string             `bson:"user_id" json:"user_id"`
	Items               []OrderItem        `bson:"items" json:"items"`
	Weight              float64            `bson:"package" json:"package"`
	TotalAmount         float64            `bson:"total_amount" json:"total_amount"`
	Status              OrderStatus        `bson:"status" json:"status"`
	ShippingMethod      string             `bson:"shipping_method" json:"shipping_method"`
	ShippingStatus      string             `bson:"shipping_status" json:"shipping_status"`
	ShippingAddress     Address            `bson:"shipping_address" json:"shipping_address"`
	ShippedTrackID      string             `bson:"shipped_track_id,omitempty" json:"shipped_track_id,omitempty"`
	PaymentURL          string             `bson:"payment_url,omitempty" json:"payment_url,omitempty"`
	ProcessedPaymentIDs []string           `bson:"processed_payment_ids,omitempty" json:"processed_payment_ids,omitempty"`
	CreatedAt           time.Time          `bson:"created_at" json:"created_at"`
	UpdatedAt           time.Time          `bson:"updated_at" json:"updated_at"`
}

// Validate validates order data
func (o *Order) Validate() error {
	if o.UserID == "" {
		return errors.New("user ID is required")
	}
	if len(o.Items) == 0 {
		return errors.New("at least one item is required")
	}
	if o.ShippingMethod == "" {
		return errors.New("Shipping method required")
	}
	if o.ShippingAddress.Street == "" {
		return errors.New("shipping address is required")
	}
	if o.ShippingAddress.Number == "" {
		return errors.New("shipping address number is required")
	}
	if o.ShippingAddress.Floor == "" {
		return errors.New("shipping address floor is required")
	}
	if o.ShippingAddress.Apartment == "" {
		return errors.New("shipping address apartment is required")
	}
	if o.ShippingAddress.City == "" {
		return errors.New("shipping address city is required")
	}
	if o.ShippingAddress.Province == "" {
		return errors.New("shipping address province is required")
	}
	if o.ShippingAddress.Country == "" {
		return errors.New("shipping address country is required")
	}
	if o.ShippingAddress.ZipCode == "" {
		return errors.New("shipping address zip code is required")
	}
	return nil
}

// CalculateTotal calculates the total amount from items
func (o *Order) CalculateTotal() {
	total := 0.0
	for _, item := range o.Items {
		total += item.Price * float64(item.Quantity)
	}
	o.TotalAmount = total
}

type UserInfo struct {
	FirstName string  `json:"first_name"`
	LastName  string  `json:"last_name"`
	DNI       string  `json:"dni"`
	Email     string  `json:"email"`
	Phone     string  `json:"phone"`
	Address   Address `json:"address"`
}
